### PR TITLE
Java 8 - 11 Blog Post

### PR DESCRIPTION
**Pull request by dferguson992**
Focuses more on the development of Java as a language over the last 5 years rather than the specific outcomes of that language.  Some of the topics mentioned herein are blog posts all their own that I would love to tackle specifically.  Regardless, I think it's worth discussing how Java has changed to adapt to a functional programming world.

### CODE REVIEW

File: posts/comparing-java-lts-releases.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+```java
+java HelloWorld.java
+```
+If you want to test something out quickly, you could mock up your application using `jshell`.  Then you can take the commands you painstaikingly tested and copy/paste them into a `poc.java` file.  This can be attached into an email for your co-worker and forwarded to them.  They can then run the application in one line and, assuming you are using native libraries only, the application will run perfectly.  This is truly a bign step for Java; Oracle has worked tirelessly to make Java a language that can stand the test of the modern programming paradigm.
```
Comment (aseigneurin) [inline posts/comparing-java-lts-releases.md]:
```suggestion
If you want to test something out quickly, you could mock up your application using `jshell`.  Then you can take the commands you painstakingly tested and copy/paste them into a `poc.java` file.  This can be attached into an email for your co-worker and forwarded to them.  They can then run the application in one line and, assuming you are using native libraries only, the application will run perfectly.  This is truly a big step for Java; Oracle has worked tirelessly to make Java a language that can stand the test of the modern programming paradigm.  
```

Comment (aseigneurin) [reply posts/comparing-java-lts-releases.md]: "Oracle has worked tirelessly" -> well, there are lots of people working on the JDK, not just Oracle

File: posts/comparing-java-lts-releases.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+```java
+java HelloWorld.java
+```
+If you want to test something out quickly, you could mock up your application using `jshell`.  Then you can take the commands you painstaikingly tested and copy/paste them into a `poc.java` file.  This can be attached into an email for your co-worker and forwarded to them.  They can then run the application in one line and, assuming you are using native libraries only, the application will run perfectly.  This is truly a bign step for Java; Oracle has worked tirelessly to make Java a language that can stand the test of the modern programming paradigm.  
+
+Languages like Python have always been used to write enterprise level applications, just like Java and C++.  However, Python has always had an edge in that it can be used in a highly functional, almost ad-hoc way.  Until Java 9, this was impossible.  Now with Java 11 introducing single-file compilation under an LTS release with lambda functions that have locally scoped, dynamically typed variables, companies can almost replace their Python scripts with Java scripts.  This is a huge win for Java.  Take for example the machine learning world.
```
Comment (aseigneurin) [inline posts/comparing-java-lts-releases.md]: dynamically typed variables -> inferred-type variables

File: posts/comparing-java-lts-releases.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
...rent results.  How does a language like Java stand a chance in a world where the function of the object outweights the object itself?  The short answer is, Java needed to be reinvented.  Now we have a robust object-oriented language that can easily fill the role of a functional scripting language.  By trimming the fat inherent to every Java application, Java once again becomes an accessible language.  We no longer need 20 lines of import statements, we can just make a module.  We no longer need to compile thousands of lines of code to see if one crucial piece works as expected, we can just use `jshell`.  We no longer have long-winded anonymous classes that define highly specific types (yes I am talking about you Apache HTTP Client), we have lambda expressions and dynamic types.  We no longer have helper classes designed to troll through your data structures for entries that meet specific criteria, we have the Streams API.  Pretty soon, [the iconic POJO may be replaced with a concept called a `record` in Java 12](https://dzone.com/articles/interesting-jdk-12-features-to-watch-out-for).  This initiaive by Oracle to keep Java modern ensures it will be used for many more years to come.
```
Comment (aseigneurin) [inline posts/comparing-java-lts-releases.md]:
```suggestion
Java has had quite the glow-up in the last 5 years.  Oracle took a language that was on the cusp of becoming one of the "old-school" languages like C or Fortran, and re-made it in the image of languages that are succeeding today for modern use-cases.  The great thing about Java used to be it was easy to learn.  There are no function pointers, there is only one type of casting, no multiple inheritance, you could run Java on any platform, and you do not need to know how to use any command line tools because your IDE takes care of most of it.  If you are a novice, the command line in your Eclipse IDE is how you interact with the application.  If you were curious, you built a JApplet.  If you were an enterprise developer, you built a REST interface underneath some HTML you copy/pasted off the web.  This worked for awhile, until our data sets got bigger, our databases all became columnar, our front-ends were stand-alone products like Grafana, and we became more interested in our future results over our current results.  How does a language like Java stand a chance in a world where the function of the object outweighs the object itself?  The short answer is, Java needed to be reinvented.  Now we have a robust object-oriented language that can easily fill the role of a functional scripting language.  By trimming the fat inherent to every Java application, Java once again becomes an accessible language.  We no longer need 20 lines of import statements, we can just make a module.  We no longer need to compile thousands of lines of code to see if one crucial piece works as expected, we can just use `jshell`.  We no longer have long-winded anonymous classes that define highly specific types (yes I am talking about you Apache HTTP Client), we have lambda expressions and dynamic types.  We no longer have helper classes designed to troll through your data structures for entries that meet specific criteria, we have the Streams API.  Pretty soon, [the iconic POJO may be replaced with a concept called a `record` in Java 12](https://dzone.com/articles/interesting-jdk-12-features-to-watch-out-for).  This initiative by Oracle to keep Java modern ensures it will be used for many more years to come.  
```

Comment (aseigneurin) [reply posts/comparing-java-lts-releases.md]: "and dynamic types" -> type inference

File: posts/comparing-java-lts-releases.md (hunk 1 above)
Comment (dferguson992) [reply posts/comparing-java-lts-releases.md]: Yeah, but I specify Oracle as the driving force because the lifecycle and design changes started as a result of Sun's acquisition by Oracle in 2010.  Java 7 released in 2011 was likely already in the works before the acquisition was finished.  Three years after Java 7 was released, Oracle released Java 8, which is pretty impressive for a company to do, making such strong language changes like lambda expressions and functional interfaces in 3 years.

Either way, the Oracle bit is ancillary and doesn't add much to the primary point of the post.

File: posts/comparing-java-lts-releases.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+authors:
+- Dan Ferguson
+tags:
+- Java
+date: 2019-01-14T14:06:00.000Z
+title: "Comparing Java LTS Releases"
+image: https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2018/01/MVP-1.png
+---
+
+# Introduction
```
Comment (aseigneurin) [inline posts/comparing-java-lts-releases.md]: Still needs to be removed ;)

File: posts/comparing-java-lts-releases.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+The big language feature introduced in Java 10 is the `var` keyword.  The `var` keyword intoduces implicitly typed variables to Java.  The concept of the implicit type has been around forever; Python and Perl use implicit typing out of the box, as does C# (with reflection or the `var` keyword).  Java has always traditionally been a statically typed language; the closest to dynamic typing with Java you could get before version 10 was with Generic classes and runtime casting.  With the introduction of the `var` keyword though, we can infer an object's type at compile time, making code easier to read.  Java is still a statically typed language, nothing will change that.  But the introduciton of the `var` keyword extends the usability of Java to use cases where developers may choose a dynamically typed language by default.  This is an exciting potential use-case for Java.  The use of the `var` keyword, combined with tools like the `jshell`, propel Java into the functional programming space.  That being said, I would caution excited fans of inferred types to use `var` only when necessary.  The following is a perfect example:
```
Comment (aseigneurin) [inline posts/comparing-java-lts-releases.md]: "the closest to dynamic typing with Java" -> I would rephrase that because Java didn't get closer to dynamic typing with the "var" keyword. Dynamic typing is when typing is done at runtime. Type inference is done at compile time, there is no type resolution at runtime.

File: posts/comparing-java-lts-releases.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
...er our current results.  How does a language like Java stand a chance in a world where the function of the object outweighs the object itself?  The short answer is, Java needed to be reinvented.  Now we have a robust object-oriented language that can easily fill the role of a functional scripting language.  By trimming the fat inherent to every Java application, Java once again becomes an accessible language.  We no longer need 20 lines of import statements, we can just make a module.  We no longer need to compile thousands of lines of code to see if one crucial piece works as expected, we can just use `jshell`.  We no longer have long-winded anonymous classes that define highly specific types (yes I am talking about you Apache HTTP Client), we have lambda expressions and dynamic types.  We no longer have helper classes designed to troll through your data structures for entries that meet specific criteria, we have the Streams API.  Pretty soon, [the iconic POJO may be replaced with a concept called a `record` in Java 12](https://dzone.com/articles/interesting-jdk-12-features-to-watch-out-for).  This initiative to keep Java modern ensures it will be used for many more years to come.
```
Comment (aseigneurin) [inline posts/comparing-java-lts-releases.md]: "lambda expressions and dynamic types" -> not dynamic :)

File: posts/comparing-java-lts-releases.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+Additionally, Java 10 improved upon Java 9's new Garbage Collector, the Garbage-First Garbage Collector.  Java 9's G1GC changed the way Java performed garbage collecting by partitioning its heap into multiple small heaps, whereas traditionally the heap would be split into three chunks.  G1GC also introduced multi-threaded processing on the garbage collector for all events save a "stop-the world," a point in time during garbage collection where program execution halts and gives the garbage collector time to assess the state of the application.  "Stop-the-world" events are unaviodable in garbage collection.  Java 10 took the multi-threaded approach perfected by Java 9 and applied it to the "stop-the-world" event, thus making the G1GC fully concurrent.  Even though the G1GC is fully concurrent and therefore faster than prior garbage collectors on average, it is worth noting it takes a larger memory footprint because of the way it chops the heap into more chunks.
```
Comment (aseigneurin) [inline posts/comparing-java-lts-releases.md]:
```suggestion
Additionally, Java 10 improved upon Java 9's new Garbage Collector, the Garbage-First Garbage Collector.  Java 9's G1GC changed the way Java performed garbage collecting by partitioning its heap into multiple small heaps, whereas traditionally the heap would be split into three chunks.  G1GC also introduced multi-threaded processing on the garbage collector for all events save a "stop-the world," a point in time during garbage collection where program execution halts and gives the garbage collector time to assess the state of the application.  "Stop-the-world" events are unavoidable in garbage collection.  Java 10 took the multi-threaded approach perfected by Java 9 and applied it to the "stop-the-world" event, thus making the G1GC fully concurrent.  Even though the G1GC is fully concurrent and therefore faster than prior garbage collectors on average, it is worth noting it takes a larger memory footprint because of the way it chops the heap into more chunks.
```

File: posts/comparing-java-lts-releases.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+Streams are Collections that have been modified for functional use.  Consider the List collection, typically a LinkedList or an ArrayList data structure.  These data structures were designed for traditional CRUD operations in a very object-oriented way.  But what happens when you need to iterate over all of elements in a list that meet a criteria, perform some chain of operations against those elements, and forward those elements on to a new list object?  You could spend a lot of time working on a "helper object" that is designed specifically to modify all of your list objects in your applciation to accomplish this, or you could use stream operations.  Streams have been integrated into many of the native Java data structures (just like many lambda functions) allowing them to be easily pulled from pre-Java 8 code.  Consider the following example pulled from [this excellent resource on the Streams API](https://winterbe.com/posts/2014/07/31/java8-stream-tutorial-examples/)
```
Comment (aseigneurin) [inline posts/comparing-java-lts-releases.md]:
```suggestion
Streams are Collections that have been modified for functional use.  Consider the List collection, typically a LinkedList or an ArrayList data structure.  These data structures were designed for traditional CRUD operations in a very object-oriented way.  But what happens when you need to iterate over all of elements in a list that meet a criteria, perform some chain of operations against those elements, and forward those elements on to a new list object?  You could spend a lot of time working on a "helper object" that is designed specifically to modify all of your list objects in your application to accomplish this, or you could use stream operations.  Streams have been integrated into many of the native Java data structures (just like many lambda functions) allowing them to be easily pulled from pre-Java 8 code.  Consider the following example pulled from [this excellent resource on the Streams API](https://winterbe.com/posts/2014/07/31/java8-stream-tutorial-examples/)
```

File: posts/comparing-java-lts-releases.md (hunk 4 above)
Comment (dferguson992) [reply posts/comparing-java-lts-releases.md]: I'm not sure what you mean, do you mean remove the title?  Or the Introduction header?  Github's UI isn't making it very clear to me lol

Comment (aseigneurin) [reply posts/comparing-java-lts-releases.md]: Remove the title (`# Introduction`), `title: "Comparing Java LTS Releases"` will be used for that instead
